src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -752,7 +760,13 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
                         break;
                 }
 
-                Debug.Assert((object)this._type != null);
+                if ((object)this._type == null)
+                {
+                    diagnostics.Add(this.ForbiddenDiagnostic, Locations[0], Name);
```

Comment:
Indeed, diagnostics from this method are always discarded. I'll remove it.

This is covered by `OutVarTests.OutVarDeconstruction_02`

There isn't a mechanism to "enforce" that we only get here for broken code, because we don't have the necessary compilation diagnostics in hand. Ideas for doing that are welcome.

---
In reply to: [91397639](https://github.com/dotnet/roslyn/pull/15548#discussion_r91397639) [](ancestors = 91397639,91395620)